src/source.rs

```diff
@@ -0,0 +1,64 @@
+use crate::{config::NatsConfig, event::NatsEvent};
+use anyhow::Result;
+use async_std::channel::{self, Sender};
+use async_std::task::spawn;
+use async_trait::async_trait;
+use fluvio::Offset;
+//use fluvio::{dataplane::record::{Record, RecordData}, Offset, RecordKey};
+use fluvio_connector_common::tracing::info;
+use fluvio_connector_common::Source;
+use futures::{stream::LocalBoxStream, StreamExt};
+use nats::asynk::Subscription as NatsSubscription;
+
+const CHANNEL_BUFFER_SIZE: usize = 10000;
+
+pub(crate) struct NatsSource {
+    host: String,
+    subject: String,
+}
+
+impl NatsSource {
+    pub(crate) fn new(config: &NatsConfig) -> Result<Self> {
+        let host = config.host.clone();
+        let subject = config.subject.clone();
+        Ok(Self { host, subject })
+    }
+}
+
+#[async_trait]
+impl<'a> Source<'a, String> for NatsSource {
+    async fn connect(self, _offset: Option<Offset>) -> Result<LocalBoxStream<'a, String>> {
+        info!("Nats host: {} subject {}", &self.host, &self.subject);
+
+        let nats_client = nats::asynk::connect(self.host).await?;
+        let nats_subscription = nats_client.subscribe(&self.subject).await?;
+
+        info!(
+            "Nats client_id {} ip {}",
+            nats_client.client_id(),
+            nats_client.client_ip().unwrap()
+        );
+
+        let (sender, receiver) = channel::bounded(CHANNEL_BUFFER_SIZE);
+        spawn(nats_loop(sender, nats_client, nats_subscription));
+        Ok(receiver.boxed_local())
+    }
+}
+
+use nats::asynk::Connection as NatsConnection;
+
+async fn nats_loop(
+    tx: Sender<String>,
+    _nats_client: NatsConnection,
+    nats_sub: NatsSubscription,
+) -> Result<()> {
+    info!("Nats loop started");
+
+    loop {
```

Comment:
Why do we need 2 loops here? If the subscription returned `None` it won't ever return `Some` in the existing code, so we get endless loop that will burn our cpu.

Reply:
I wondered the same - e.g. see [kafka-source](https://github.com/infinyon/kafka-connector/blob/main/crates/kafka-source/src/source.rs#L57) - mqtt-source at least has [some error handling](https://github.com/infinyon/mqtt-connector/blob/main/src/source.rs#L106) - uniform more ergonomic approach here would be nice

Reply:
I moved the connection stuff inside first loop so it reconnects.

Reply:
Still not clear what we are actually retrying here. All errors will lead to fast return from the method.  Only in case of `nats_subscription.next().await` returns `None` we do retry. We do not reconnect if, for example, message deserialization fails.

Reply:
The `None` returns in case connection was closed and that is most common realistically to happen thus should be handled. If the serialization fails that is a bug since serialization is a code problem in the connector ? The problem also becomes how do we signal this to cluster - the only mechanism to handle failures like this has been just panic - probably something that needs to have a pattern that can be used - not something we can solve in this PR I believe. The reconnect needs to be handled and this is now handled in the loop.

In case I missed something please let me know

Reply:
Yeah, looks the nats client has no way of indicating IO errors during message reading. So if it reads a message, it reads it completely, in that case, I agree that the deserialization issue is a developer mistake, not business logic. It also means that `nats_subscription.next() returns None` is the only indication of a failed connection, and our retry covers it, so, it is ok. Then, yes, looks good to me.